nutils/evaluable.py

```diff
@@ -757,10 +757,31 @@ def _assparse(self):
       return chunks
     return _assparse
 
+  def _evalf_checker(orig):
+    bind_orig = getattr(orig, '__get__', lambda *args: orig)
+    class method_wrapper:
+      def __get__(self, instance, owner):
+        bound_orig = bind_orig(instance, owner)
+        @functools.wraps(orig)
+        def bound_method_wrapper(*args, **kwargs):
+          res = bound_orig(*args, **kwargs)
+          if hasattr(instance, 'dtype'):
+            assert numpy.can_cast(res.dtype, instance.dtype, casting='safe'), ((instance.dtype, res.dtype), instance, res)
+          try:
+            t = Tuple(instance.shape).eval()
```

Comment:
All items of `instance.shape` for which this evaluation works are already integers (see `as_canonical_length`) so you can loop over them and test only those. It's probably also useful to test `ndim` separately.